js-pkg/packages/driftdb/src/webrtc.ts

```diff
@@ -0,0 +1,292 @@
+  public connMap = new Map<string, ChannelSendReceive>()
+  private onMessage = (msg: DataChannelMsg) => {
+    console.log('unhandled', msg)
+  }
+  private onFailure = (conn: RTCPeerConnection) => {
+    conn.restartIce()
+  }
+
+  constructor(private db: DbConnection, public myId: string, private throttleMs = 0) {}
+
+  addNewConnection(p2: string) {
+    const signalingChannel = new SignalingChannel(this.db, this.myId, p2)
+    const conn = signalingChannel.createWebRTCConnection(
+      [getDataChannelCreator(this.myId, p2, (msg) => this.onMessage(msg))],
+      this.onFailure
+    )[0]
+    return conn
+  }
+
+  getConn(peer: string) {
+    return this.connMap.get(this.myId + peer)
+  }
+
+  setConnMap(entries: [string, ChannelSendReceive][]) {
+    this.connMap = new Map(entries.map(([peer, conn]) => [this.myId + peer, conn]))
+  }
+
+  send = throttle((msg: string) => {
+    this.connMap.forEach((peer) => {
+      peer.send(msg)
+    })
+  }, this.throttleMs)
+
+  setOnMessage(func: OnMessage) {
+    this.onMessage = func
+    this.connMap.forEach((peer) => {
+      peer.onMessage = func
+    })
+  }
+
+  setOnFailure(func: (conn: RTCPeerConnection) => void) {
+    this.onFailure = func
+  }
+}
+
+export class SyncedWebRTCConnections extends WebRTCConnections {
+  presence: PresenceListener<string>
+  private peersToLastMsg: Record<string, WrappedPresenceMessage<any>> = {}
+
+  constructor(db: DbConnection, id: string, throttle = 0) {
+    super(db, id, throttle)
+    this.presence = new PresenceListener<string>({
+      initialState: '',
+      db,
+      clientId: id,
+      callback: (msg) => {
+        this.sync(Object.values(msg).map(({ value }) => value))
+      },
+      minPresenceInterval: throttle
+    })
+    this.refreshConnections()
+    this.presence.subscribe()
+    this.setOnMessage((_msg) => {})
+    this.setOnFailure(this.createFailureHandler())
+  }
+
+  createFailureHandler() {
+    let numErrors = 0
+    return (conn: RTCPeerConnection) => {
+      numErrors++
+      if (numErrors == 3) {
+        this.refreshConnections()
+      } else if (numErrors > 3) {
+        conn.close()
+      } else {
+        conn.restartIce()
+      }
+    }
+  }
+
+  setOnMessage(func: OnMessage) {
+    super.setOnMessage((msg) => {
+      func(msg)
+      this.peersToLastMsg[msg.sender] = msg
+    })
+  }
+
+  refreshConnections() {
+    let newId = typeof crypto !== 'undefined' ? crypto.randomUUID().slice(0, 5) : 'PLACEHOLDER'
+    this.myId = newId
+    this.presence.updateState(newId)
+    this.setOnFailure(this.createFailureHandler())
+  }
+
+  sync(newPeers: string[]) {
+    this.setConnMap(
+      newPeers.map((peer) => [peer, this.getConn(peer) ?? this.addNewConnection(peer)])
+    )
+    this.peersToLastMsg = Object.fromEntries(
+      Object.entries(this.peersToLastMsg).reduce(
+        (a, b) => (newPeers.includes(b[0]) ? [...a, b] : [...a]),
+        [] as [string, WrappedPresenceMessage<any>][]
+      )
+    )
```

Comment:
Minor comment: there might be a simpler (and more performant!) way to write this.

Three things about this:

1. Creates a lot of unnecessary objects that are just thrown away by the end of the fn (an array when converting newPeersToLastMsg, an array for the reduce init value, and an array for every key in peersToLastMsg)
2. Uses includes() which isn't es2015
3. Is just hard to read 😬 

Consider this option, which is far more boring, and also doesn't suffer from the above:

```ts
const newPeersToLastMsg: Record<string, WrappedPresenceMessage<any>> = {}
for (const peer of newPeers) {
  newPeersToLastMsg[peer] = this.peersToLastMsg[peer]
}
this.peersToLastMsg = newPeersToLastMsg
```

Or this, which reuses the same object over and over:

```ts
for (const peer in this.peersToLastMsg) {
  if (!this.connMap.has(peer)) {
    delete this.peersToLastMsg[peer]
  }
}
```